Throwing an exception crashes a program on FreeBSD because typeinfo for BasicError<char> isn't exported

**Describe the bug**
GNOME Photos crashes at launch because one of its dependency, gexiv2, calls an Exiv2 function which throws an exception.

**To Reproduce**
The example program posted at https://github.com/Exiv2/exiv2/issues/644#issuecomment-452472029 is probably the most simple way to reproduce the problem. It seems to come from [`gexiv2_metadata_register_xmp_namespace` function of gexiv2](https://gitlab.gnome.org/GNOME/gexiv2/blob/gexiv2-0.12.0/gexiv2/gexiv2-metadata-xmp.cpp#L402) and it is where GNOME Photos crashes.

**Expected behavior**
No crash should happen because the exception is caught.

**Desktop:**
 - Exiv2 0.27.1
 - FreeBSD 12.0
 - Clang 6.0.1

**Additional context**
I am not sure if I understand the problem correctly, so I just post all of my findings here. This probably makes the issue description longer than it needs, but I hope the information will be helpful for you to understand the problem if you aren't going to test Exiv2 on FreeBSD.

This is what I saw when running GNOME Photos. It just crashes without printing any message.
```
$ gnome-photos
Abort trap (core dumped)
$ gdb gnome-photos gnome-photos.core
...
Core was generated by `gnome-photos'.
Program terminated with signal SIGABRT, Aborted.
(gdb) bt
#0  thr_kill () at thr_kill.S:3
#1  0x0000000801b68844 in __raise (s=6) at /usr/src/lib/libc/gen/raise.c:52
#2  0x0000000801adb079 in abort () at /usr/src/lib/libc/stdlib/abort.c:67
#3  0x00000008023d98ba in report_failure (err=<optimized out>, thrown_exception=0x80aefcc08) at /usr/src/contrib/libcxxrt/exception.cc:719
#4  0x000000080224fc01 in Exiv2::XmpProperties::nsInfoUnsafe (prefix="gnome") at /home/lantw44/gnome/source/exiv2/src/properties.cpp:2659
#5  0x000000080224fafb in Exiv2::XmpProperties::ns (prefix="gnome") at /home/lantw44/gnome/source/exiv2/src/properties.cpp:2586
#6  0x00000008004fcad2 in gexiv2_metadata_register_xmp_namespace (name=0x209d5a "http://www.gnome.org/xmp", prefix=0x20d1d6 "gnome") at ../../source/gexiv2/gexiv2/gexiv2-metadata-xmp.cpp:407
#7  0x000000000028168b in photos_application_create_window (self=0x8043381f0) at photos-application.c:813
#8  0x000000000027ff22 in photos_application_activate (application=<optimized out>) at photos-application.c:2436
#9  0x000000080172738d in _g_closure_invoke_va (closure=0x8009ff8e0, return_value=0x0, instance=0x8043381f0, args=<optimized out>, n_params=0, param_types=0x0) at ../../source/glib/gobject/gclosure.c:873
#10 0x000000080173b23b in g_signal_emit_valist (instance=0x8043381f0, signal_id=<optimized out>, detail=0, var_args=0x7fffffffd430) at ../../source/glib/gobject/gsignal.c:3306
#11 0x000000080173bba4 in g_signal_emit (instance=0x188ea, signal_id=6, detail=0) at ../../source/glib/gobject/gsignal.c:3453
#12 0x000000000027ffd8 in photos_application_command_line (application=0x8043381f0, command_line=0x8076e3340) at photos-application.c:2468
#13 0x000000080163dcfc in _g_cclosure_marshal_INT__OBJECTv (closure=0x8009ff920, return_value=0x7fffffffd550, instance=0x8043381f0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, 
    param_types=0x80033e0e8) at ../../source/glib/gio/gmarshal-internal.c:845
#14 0x000000080172738d in _g_closure_invoke_va (closure=0x8009ff920, return_value=0x7fffffffd550, instance=0x8043381f0, args=<optimized out>, n_params=1, param_types=0x80033e0e8)
    at ../../source/glib/gobject/gclosure.c:873
#15 0x000000080173b23b in g_signal_emit_valist (instance=0x8043381f0, signal_id=<optimized out>, detail=0, var_args=0x7fffffffd740) at ../../source/glib/gobject/gsignal.c:3306
#16 0x000000080173bba4 in g_signal_emit (instance=0x188ea, signal_id=6, detail=0) at ../../source/glib/gobject/gsignal.c:3453
#17 0x000000080168fb17 in g_application_call_command_line (application=0x8043381f0, arguments=<optimized out>, options=<optimized out>, exit_status=0x7fffffffd814) at ../../source/glib/gio/gapplication.c:1076
#18 0x00000008016911cb in g_application_real_local_command_line (application=0x8043381f0, arguments=<optimized out>, exit_status=0x7fffffffd814) at ../../source/glib/gio/gapplication.c:1130
#19 0x000000080168f8b5 in g_application_run (application=0x8043381f0, argc=1, argv=0x7fffffffd8f8) at ../../source/glib/gio/gapplication.c:2523
#20 0x00000000002da87a in main (argc=1, argv=0x7fffffffd8f8) at photos-main.c:61
```

I quickly found a way to avoid the crash based on commit 04c9b181cc2c7741fcb134ccc43f2bafc1f86f19.
```diff
diff --git a/cmake/mainSetup.cmake b/cmake/mainSetup.cmake
index a4d983d4..1de6324c 100644
--- a/cmake/mainSetup.cmake
+++ b/cmake/mainSetup.cmake
@@ -13,7 +13,9 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
 
-if (NOT CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
+if (NOT ((CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang") OR
+         (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND
+          CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")))
     set(CMAKE_CXX_VISIBILITY_PRESET hidden)
     set(CMAKE_VISIBILITY_INLINES_HIDDEN 1)
 endif()
```
But this isn't a solution. It is just a workaround because it effectively disables visibility control.

In src/properties.cpp, function `XmpProperties::nsInfoUnsafe`:
```cpp
const XmpNsInfo* XmpProperties::nsInfoUnsafe(const std::string& prefix)
{
    const XmpNsInfo::Prefix pf(prefix);
    const XmpNsInfo* xn = lookupNsRegistryUnsafe(pf);
    if (!xn) xn = find(xmpNsInfo, pf);
    if (!xn) throw Error(kerNoNamespaceInfoForXmpPrefix, prefix);
    return xn;
}
```
It was the `throw` line that crashed the program. By using GDB, I found the execution of `throw` began at an internal function [`__cxa_throw` defined in /usr/src/contrib/libcxxrt/exception.cc](https://svnweb.freebsd.org/base/releng/12.0/contrib/libcxxrt/exception.cc?view=markup#l782). It then called [`throw_exception` defined in the same file](https://svnweb.freebsd.org/base/releng/12.0/contrib/libcxxrt/exception.cc?view=markup#l759) to continue the process.
```cpp
static void throw_exception(__cxa_exception *ex)
{
    // [snip]

    _Unwind_Reason_Code err = _Unwind_RaiseException(&ex->unwindHeader);
    // The _Unwind_RaiseException() function should not return, it should
    // unwind the stack past this function.  If it does return, then something
    // has gone wrong.
    report_failure(err, ex);
}
```
This was where `report_failure` in the crash dump came from. GDB told me `err` was `_URC_END_OF_STACK`, which suggested that libunwind couldn't find the function catching the exception.

There is an environment variable called `LIBUNWIND_PRINT_UNWINDING` which can be used to debug libunwind. In a working build, it successfully found the function catching the exception was `gexiv2_metadata_register_xmp_namespace`:
```
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): pc=0x802466416, start_ip=0x802466370, func=__cxa_throw, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): pc=0x8022cff61, start_ip=0x8022cfed0, func=_ZN5Exiv213XmpProperties12nsInfoUnsafeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE, lsda=0x8020824e4, personality=0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): calling personality function 0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): _URC_CONTINUE_UNWIND
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): pc=0x8022cfe5b, start_ip=0x8022cfde0, func=_ZN5Exiv213XmpProperties2nsERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE, lsda=0x802082470, personality=0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): calling personality function 0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): _URC_CONTINUE_UNWIND
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): pc=0x8004fcad2, start_ip=0x8004fca80, func=gexiv2_metadata_register_xmp_namespace, lsda=0x8004e2d98, personality=0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): calling personality function 0x802466a60
libunwind: unwind_phase1(ex_ojb=0x80b08fa60): _URC_HANDLER_FOUND
```

In a broken build, it couldn't find the function. The function `gexiv2_metadata_register_xmp_namespace` existed in the call stack, but it seemed to ignore   it.
```
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x8023d9416, start_ip=0x8023d9370, func=__cxa_throw, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80224fc01, start_ip=0x80224fb70, func=_ZN5Exiv213XmpProperties12nsInfoUnsafeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE, lsda=0x8020824e4, personality=0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): calling personality function 0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): _URC_CONTINUE_UNWIND
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80224fafb, start_ip=0x80224fa80, func=_ZN5Exiv213XmpProperties2nsERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE, lsda=0x802082470, personality=0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): calling personality function 0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): _URC_CONTINUE_UNWIND
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x8004fcad2, start_ip=0x8004fca80, func=gexiv2_metadata_register_xmp_namespace, lsda=0x8004e2d98, personality=0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): calling personality function 0x8023d9a60
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): _URC_CONTINUE_UNWIND
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x28168b, start_ip=0x281640, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x27ff22, start_ip=0x27fef0, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80172738d, start_ip=0x8017272c0, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80173b23b, start_ip=0x80173aec0, func=g_signal_emit_valist, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80173bba4, start_ip=0x80173bb30, func=g_signal_emit, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x27ffd8, start_ip=0x27ff70, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80163dcfc, start_ip=0x80163dc70, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80172738d, start_ip=0x8017272c0, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80173b23b, start_ip=0x80173aec0, func=g_signal_emit_valist, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80173bba4, start_ip=0x80173bb30, func=g_signal_emit, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80168fb17, start_ip=0x80168faa0, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x8016911cb, start_ip=0x801691080, func=.anonymous., lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x80168f8b5, start_ip=0x80168f790, func=g_application_run, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x2da87a, start_ip=0x2da820, func=main, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): pc=0x27f11b, start_ip=0x27f000, func=_start, lsda=0x0, personality=0x0
libunwind: unwind_phase1(ex_ojb=0x80bb5ce60): unw_step() reached bottom => _URC_END_OF_STACK
````

Therefore, I followed the execution of `_Unwind_RaiseException` in order to find why the visibility control broke it. [`_Unwind_RaiseException` is defined in /usr/src/contrib/llvm/projects/libunwind/src/UnwindLevel1.c](https://svnweb.freebsd.org/base/releng/12.0/contrib/llvm/projects/libunwind/src/UnwindLevel1.c?view=markup#l340). It then called [`unwind_phase1` defined in the same file](https://svnweb.freebsd.org/base/releng/12.0/contrib/llvm/projects/libunwind/src/UnwindLevel1.c?view=markup#l35) to search the stack.
```cpp
    // If there is a personality routine, ask it if it will want to stop at
    // this frame.
    if (frameInfo.handler != 0) {
      __personality_routine p =
          (__personality_routine)(long)(frameInfo.handler);
      _LIBUNWIND_TRACE_UNWINDING(
          "unwind_phase1(ex_ojb=%p): calling personality function %p",
          (void *)exception_object, (void *)(uintptr_t)p);
      _Unwind_Reason_Code personalityResult =
          (*p)(1, _UA_SEARCH_PHASE, exception_object->exception_class,
               exception_object, (struct _Unwind_Context *)(cursor));
      switch (personalityResult) {
      case _URC_HANDLER_FOUND:
        // [snip] 

      case _URC_CONTINUE_UNWIND:
        // [snip]
    }
```
The key is the value of `personalityResult` here. In a working build, `_URC_HANDLER_FOUND` stopped the search. In a broken build, it was always `_URC_CONTINUE_UNWIND` until it reached the end of the stack. `personalityResult` was the return value of function pointer `p`, and `p` was [`__gxx_personality_v0` defined in /usr/src/contrib/libcxxrt/exception.cc](https://svnweb.freebsd.org/base/releng/12.0/contrib/libcxxrt/exception.cc?view=markup#l1071).

It turned out `_URC_HANDLER_FOUND` was returned from here in a working build:
```cpp
        handler_type found_handler = check_action_record(context, &lsda,
                action.action_record, realEx, &selector, ex->adjustedPtr);
        // If there's no action record, we've only found a cleanup, so keep
        // searching for something real
        if (found_handler == handler_catch)
        {
            // Cache the results for the phase 2 unwind, if we found a handler
            // and this is not a foreign exception.
            if (ex)
            {
                saveLandingPad(context, exceptionObject, ex, selector, action.landing_pad);
                ex->languageSpecificData = reinterpret_cast<const char*>(lsda_addr);
                ex->actionRecord = reinterpret_cast<const char*>(action.action_record);
                // ex->adjustedPtr is set when finding the action record.
            }
            return _URC_HANDLER_FOUND;
        }
```
[`check_action_record` is defined in the same file](https://svnweb.freebsd.org/base/releng/12.0/contrib/libcxxrt/exception.cc?view=markup#l971):
```cpp
        // We only check handler types for C++ exceptions - foreign exceptions
        // are only allowed for cleanups and catchalls.
        if (filter > 0)
        {
            std::type_info *handler_type = get_type_info_entry(context, lsda, filter);
            if (check_type_signature(ex, handler_type, adjustedPtr))
            {
                *selector = filter;
                return handler_catch;
            }
        }
```
`handler_type` was `typeinfo for Exiv2::BasicError<char>` in both working and broken build. The difference was the return value of [`check_type_signature`](https://svnweb.freebsd.org/base/releng/12.0/contrib/libcxxrt/exception.cc?view=markup#l923). In a working build, it returned `true`. In a broken build, it returned `false`.
```cpp
static bool check_type_signature(__cxa_exception *ex,
                                 const std::type_info *type,
                                 void *&adjustedPtr)
{
    // [snip]

    // If the types are the same, no casting is needed.
    if (*type == *ex_type)
    {
        adjustedPtr = exception_ptr;
        return true; // <- It should return from here.
    }


    if (type->__do_catch(ex_type, &exception_ptr, 1))
    {
        adjustedPtr = exception_ptr;
        return true;
    }

    return false; // <- But a broken build returned from here.
}
```
The result suggested it was the visibility problem of `typeinfo for Exiv2::BasicError<char>` that caused the crash. In a broken build, `typeinfo for Exiv2::BasicError<char>` couldn't be found in the output of `nm -DC`. In a working build, it was there and it was marked as `V`, a weak object.

Finally, I got a working build with this patch:
```diff
diff --git a/include/exiv2/error.hpp b/include/exiv2/error.hpp
index 42a0959c..bd74c449 100644
--- a/include/exiv2/error.hpp
+++ b/include/exiv2/error.hpp
@@ -260,7 +260,7 @@ namespace Exiv2 {
              provided to print errors to a stream.
      */
     template<typename charT>
-    class BasicError : public AnyError {
+    class EXIV2API BasicError : public AnyError {
     public:
         //! @name Creators
         //@{
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 34087004..df041890 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -145,7 +145,7 @@ endif()
 
 set_target_properties( exiv2lib_int PROPERTIES
     POSITION_INDEPENDENT_CODE ON
-    COMPILE_DEFINITIONS exiv2lib_STATIC
+#   COMPILE_DEFINITIONS exiv2lib_STATIC
 )
 
 

```
Both changes are required for `typeinfo for Exiv2::BasicError<char>` to be exported. If I only modified `include/exiv2/error.hpp`, files belonging to `exiv2lib_int` will hide it because `EXIV2API` expands to nothing when `exiv2lib_STATIC` is defined.

It can be found by using `readelf -aW` on `.o` files and searching for `_ZTIN5Exiv210BasicErrorIcEE`, the mangled name of `typeinfo for Exiv2::BasicError<char>`. In src/CMakeFiles/exiv2.dir and src/CMakeFiles/exiv2lib.dir, they are 'WEAK DEFAULT', which is fine. However, in src/CMakeFiles/exiv2lib_int.dir, they are 'WEAK HIDDEN'. The result is that the symbol became 'LOCAL HIDDEN' in the linked shared library, preventing it from being used from other libraries or programs. If `exiv2lib_STATIC` is removed, all objects have 'WEAK DEFAULT', so the linked shared library also has 'WEAK DEFAULT'.

Here is the difference of dynamic symbols between the current upstream version and the version using the above patch.
```diff
--- broken	2019-07-06 01:43:00.523185000 +0800
+++ working	2019-07-06 12:20:05.783945000 +0800
@@ -1542,6 +1542,7 @@
 V construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_ofstream<char, std::__1::char_traits<char> >
 V construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 V construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
+V typeinfo for Exiv2::BasicError<char>
 V typeinfo for Exiv2::HttpIo
 V typeinfo for Exiv2::IoCloser
 V typeinfo for std::__1::basic_filebuf<char, std::__1::char_traits<char> >
@@ -1550,6 +1551,7 @@
 V typeinfo for std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 V typeinfo for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 V typeinfo for std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
+V typeinfo name for Exiv2::BasicError<char>
 V typeinfo name for Exiv2::HttpIo
 V typeinfo name for Exiv2::IoCloser
 V typeinfo name for std::__1::basic_filebuf<char, std::__1::char_traits<char> >
@@ -1558,6 +1560,7 @@
 V typeinfo name for std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 V typeinfo name for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 V typeinfo name for std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
+V vtable for Exiv2::BasicError<char>
 V vtable for Exiv2::HttpIo
 V vtable for Exiv2::IoCloser
 V vtable for std::__1::basic_filebuf<char, std::__1::char_traits<char> >
@@ -1572,5 +1575,44 @@
 V VTT for std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >
 w __cxa_finalize
 w _Jv_RegisterClasses
+W Exiv2::BasicError<char>::~BasicError()
+W Exiv2::BasicError<char>::~BasicError()
+W Exiv2::BasicError<char>::BasicError(Exiv2::ErrorCode)
+W Exiv2::BasicError<char>::BasicError<char [10]>(Exiv2::ErrorCode, char const (&) [10])
+W Exiv2::BasicError<char>::BasicError<char [109]>(Exiv2::ErrorCode, char const (&) [109])
+W Exiv2::BasicError<char>::BasicError<char [11]>(Exiv2::ErrorCode, char const (&) [11])
+W Exiv2::BasicError<char>::BasicError<char [13]>(Exiv2::ErrorCode, char const (&) [13])
+W Exiv2::BasicError<char>::BasicError<char [14], char [10]>(Exiv2::ErrorCode, char const (&) [14], char const (&) [10])
+W Exiv2::BasicError<char>::BasicError<char [14], char [4]>(Exiv2::ErrorCode, char const (&) [14], char const (&) [4])
+W Exiv2::BasicError<char>::BasicError<char [14], char [5]>(Exiv2::ErrorCode, char const (&) [14], char const (&) [5])
+W Exiv2::BasicError<char>::BasicError<char [15]>(Exiv2::ErrorCode, char const (&) [15])
+W Exiv2::BasicError<char>::BasicError<char [21]>(Exiv2::ErrorCode, char const (&) [21])
+W Exiv2::BasicError<char>::BasicError<char [24]>(Exiv2::ErrorCode, char const (&) [24])
+W Exiv2::BasicError<char>::BasicError<char [26]>(Exiv2::ErrorCode, char const (&) [26])
+W Exiv2::BasicError<char>::BasicError<char [37]>(Exiv2::ErrorCode, char const (&) [37])
+W Exiv2::BasicError<char>::BasicError<char [4]>(Exiv2::ErrorCode, char const (&) [4])
+W Exiv2::BasicError<char>::BasicError<char [5]>(Exiv2::ErrorCode, char const (&) [5])
+W Exiv2::BasicError<char>::BasicError<char [53]>(Exiv2::ErrorCode, char const (&) [53])
+W Exiv2::BasicError<char>::BasicError<char [7], int>(Exiv2::ErrorCode, char const (&) [7], int const&)
+W Exiv2::BasicError<char>::BasicError<char [7], long>(Exiv2::ErrorCode, char const (&) [7], long const&)
+W Exiv2::BasicError<char>::BasicError<char const*>(Exiv2::ErrorCode, char const* const&)
+W Exiv2::BasicError<char>::BasicError<char*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [7]>(Exiv2::ErrorCode, char* const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [7])
+W Exiv2::BasicError<char>::BasicError<char*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [8]>(Exiv2::ErrorCode, char* const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [8])
+W Exiv2::BasicError<char>::BasicError<char*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [9]>(Exiv2::ErrorCode, char* const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [9])
+W Exiv2::BasicError<char>::BasicError<int>(Exiv2::ErrorCode, int const&)
+W Exiv2::BasicError<char>::BasicError<long, char const*>(Exiv2::ErrorCode, long const&, char const* const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [3], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [3], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [4], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [4], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char const*>(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const* const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char* const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [13]>(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [13])
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [5]>(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [5])
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [7]>(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [7])
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
+W Exiv2::BasicError<char>::BasicError<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned long>(Exiv2::ErrorCode, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long const&)
+W Exiv2::BasicError<char>::code() const
+W Exiv2::BasicError<char>::what() const
 W Exiv2::Xmpdatum& Exiv2::Xmpdatum::operator=<char [30]>(char const (&) [30])
 W Exiv2::Xmpdatum& Exiv2::Xmpdatum::operator=<int>(int const&)
```

Do you think it is OK to export `BasicError`? I think the use of `exiv2lib_STATIC` here may be incorrect because `exiv2lib_int` gets linked into a shared library, but it isn't a library which is intended to be linked staticlly into other programs.

This isn't working on Ubuntu.  Can't link jens.cpp
```
rmills@rmillsmm-ubuntu:~/temp$ cat jens.cpp
// file: jens.cpp
// https://github.com/Exiv2/exiv2/issues/644#issuecomment-452472029

/*
605 rmills@rmillsmbp:~/temp $ make jens LDFLAGS=-lexiv2
c++   -lexiv2  jens.cpp   -o jens
606 rmills@rmillsmbp:~/temp $ ./jens
Caught error No namespace info available for XMP prefix `gnome'
http://www.gnome.org/xmp/
607 rmills@rmillsmbp:~/temp $
*/

#include <exiv2/exiv2.hpp>
#include <iostream>

int main(int argc, char *argv[])
{
        try {
                Exiv2::XmpProperties::ns("gnome");
        } catch (Exiv2::Error &error) {
                std::cerr << "Caught error " << error.what() << "\n";
                Exiv2::XmpProperties::registerNs("http://www.gnome.org/xmp", "gnome");
        }
        std::cout << Exiv2::XmpProperties::ns("gnome") << std::endl;
}
rmills@rmillsmm-ubuntu:~/temp$ make jens LDFLAGS=-lexiv2
g++   -lexiv2  jens.cpp   -o jens
/tmp/ccEaKplI.o: In function `main':
jens.cpp:(.text+0x62): undefined reference to `Exiv2::XmpProperties::ns(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
jens.cpp:(.text+0xc5): undefined reference to `Exiv2::XmpProperties::ns(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
jens.cpp:(.text+0x22f): undefined reference to `Exiv2::XmpProperties::registerNs(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/tmp/ccEaKplI.o:(.data.rel.ro._ZTIN5Exiv210BasicErrorIcEE[_ZTIN5Exiv210BasicErrorIcEE]+0x10): undefined reference to `typeinfo for Exiv2::AnyError'
collect2: error: ld returned 1 exit status
<builtin>: recipe for target 'jens' failed
make: *** [jens] Error 1
rmills@rmillsmm-ubuntu:~/temp$ 
```

So, @lantw44 has a work around for FreeBSD which also works on MacOS-X (which was the subject of #644).  I think we have to get rid of the template in `include/exiv2/error.hpp`.  

This isn't ready to ship in Exiv2 v0.27.2 RC3.  We'll have to defer this for v0.27.3 (scheduled 2019-09-30) and solve #685 as well.

--- a/include/exiv2/error.hpp
+++ b/include/exiv2/error.hpp
@@ -265,38 +265,38 @@ namespace Exiv2 {
         //! @name Creators
         //@{
         //! Constructor taking only an error code
-        explicit BasicError(ErrorCode code);
+        explicit inline BasicError(ErrorCode code);
 
         //! Constructor taking an error code and one argument
         template<typename A>
-        BasicError(ErrorCode code, const A& arg1);
+        inline BasicError(ErrorCode code, const A& arg1);
 
         //! Constructor taking an error code and two arguments
         template<typename A, typename B>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2);
 
         //! Constructor taking an error code and three arguments
         template<typename A, typename B, typename C>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
 
         //! Virtual destructor. (Needed because of throw())
-        virtual ~BasicError() throw();
+        virtual inline ~BasicError() throw();
         //@}
 
         //! @name Accessors
         //@{
-        virtual int code() const throw();
+        virtual inline int code() const throw();
         /*!
           @brief Return the error message as a C-string. The pointer returned by what()
                  is valid only as long as the BasicError object exists.
          */
-        virtual const char* what() const throw();
+        virtual inline const char* what() const throw();
 #ifdef EXV_UNICODE_PATH
         /*!
           @brief Return the error message as a wchar_t-string. The pointer returned by
                  wwhat() is valid only as long as the BasicError object exists.
          */
-        virtual const wchar_t* wwhat() const throw();
+        virtual inline const wchar_t* wwhat() const throw();
 #endif
         //@}
 
@@ -304,7 +304,7 @@ namespace Exiv2 {
         //! @name Manipulators
         //@{
         //! Assemble the error message from the arguments
-        EXIV2API void setMsg();
+        void setMsg();
         //@}
 
         // DATA

--- a/include/exiv2/error.hpp
+++ b/include/exiv2/error.hpp
@@ -260,7 +260,7 @@ namespace Exiv2 {
              provided to print errors to a stream.
      */
     template<typename charT>
-    class BasicError : public AnyError {
+    class EXIV2API BasicError : public AnyError {
     public:
         //! @name Creators
         //@{

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -145,9 +145,7 @@ endif()
 
 set_target_properties( exiv2lib_int PROPERTIES
     POSITION_INDEPENDENT_CODE ON
-    COMPILE_DEFINITIONS exiv2lib_STATIC
 )
-
 
 target_include_directories(exiv2lib_int PRIVATE ${ZLIB_INCLUDE_DIR})
 target_include_directories(exiv2lib PRIVATE ${ZLIB_INCLUDE_DIR})

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -145,6 +145,7 @@ endif()
 
 set_target_properties( exiv2lib_int PROPERTIES
     POSITION_INDEPENDENT_CODE ON
+    COMPILE_DEFINITIONS exiv2lib_EXPORTS
 )
 
 target_include_directories(exiv2lib_int PRIVATE ${ZLIB_INCLUDE_DIR})

--- a/src/error.cpp
+++ b/src/error.cpp
@@ -225,7 +225,7 @@ namespace Exiv2 {
     }
 
     template<>
-    void BasicError<char>::setMsg()
+    void EXIV2API BasicError<char>::setMsg()
     {
         std::string msg = _(errMsg(code_));
         std::string::size_type pos;
@@ -262,7 +262,7 @@ namespace Exiv2 {
 
 #ifdef EXV_UNICODE_PATH
     template<>
-    void BasicError<wchar_t>::setMsg()
+    void EXIV2API BasicError<wchar_t>::setMsg()
     {
         std::string s = _(errMsg(code_));
         std::wstring wmsg(s.begin(), s.end());

--- a/unitTests/CMakeLists.txt
+++ b/unitTests/CMakeLists.txt
@@ -34,11 +34,11 @@ add_executable(unit_tests mainTestRunner.cpp
     ${unit_tests_exiv2lib_SOURCES}
 )
 
-target_compile_definitions(unit_tests
-    PRIVATE
-        exiv2lib_STATIC
-        ${exiv2lib_COMPILE_DEFINITIONS}
-)
+target_compile_definitions(unit_tests PRIVATE exiv2lib_STATIC)
+
+if (exiv2lib_COMPILE_DEFINITIONS)
+    target_compile_definitions(unit_tests PRIVATE ${exiv2lib_COMPILE_DEFINITIONS})
+endif ()
 
 target_include_directories(unit_tests
     PRIVATE

--- a/unitTests/CMakeLists.txt
+++ b/unitTests/CMakeLists.txt
@@ -1,5 +1,22 @@
 find_package(GTest REQUIRED)
 
+get_target_property(exiv2lib_SOURCES exiv2lib SOURCES)
+get_target_property(exiv2lib_int_SOURCES exiv2lib_int SOURCES)
+get_target_property(exiv2lib_COMPILE_DEFINITIONS exiv2lib COMPILE_DEFINITIONS)
+get_target_property(exiv2lib_INCLUDE_DIRECTORIES exiv2lib INCLUDE_DIRECTORIES)
+get_target_property(exiv2lib_LINK_LIBRARIES exiv2lib LINK_LIBRARIES)
+
+set(unit_tests_exiv2lib_SOURCES)
+foreach(source IN LISTS exiv2lib_SOURCES exiv2lib_int_SOURCES)
+    if(source MATCHES "\.(c|cpp|h|hpp)$")
+        if(source MATCHES ".*/.*")
+            list(APPEND unit_tests_exiv2lib_SOURCES "${source}")
+        else()
+            list(APPEND unit_tests_exiv2lib_SOURCES "../src/${source}")
+        endif()
+    endif()
+endforeach()
+
 add_executable(unit_tests mainTestRunner.cpp
     gtestwrapper.h
     test_types.cpp
@@ -14,13 +31,24 @@ add_executable(unit_tests mainTestRunner.cpp
     test_helper_functions.cpp
     test_slice.cpp
     test_image_int.cpp
-    $<TARGET_OBJECTS:exiv2lib_int>
+    ${unit_tests_exiv2lib_SOURCES}
+)
+
+target_compile_definitions(unit_tests
+    PRIVATE
+        exiv2lib_STATIC
+        ${exiv2lib_COMPILE_DEFINITIONS}
+)
+
+target_include_directories(unit_tests
+    PRIVATE
+        ${exiv2lib_INCLUDE_DIRECTORIES}
 )
 
 #TODO Use GTest::GTest once we upgrade the minimum CMake version required
 target_link_libraries(unit_tests
     PRIVATE
-        exiv2lib
+        ${exiv2lib_LINK_LIBRARIES}
         ${GTEST_BOTH_LIBRARIES}
 )
 

